packages/peer-exchange/src/waku_peer_exchange.ts

```diff
@@ -40,50 +49,75 @@ export class WakuPeerExchange implements PeerExchange {
   async query(params: PeerExchangeQueryParams): Promise<PeerExchangeResponse> {
     const { numPeers } = params;
 
-    try {
-      const rpcQuery = PeerExchangeRPC.createRequest({ numPeers });
+    console.log({ numPeers });
 
-      const peer = await this.getPeer();
+    const rpcQuery = PeerExchangeRPC.createRequest({ numPeers });
+    console.log(rpcQuery);
 
-      const stream = await this.newStream(peer);
+    const peer = await this.getPeer();
 
-      const pipeResponse = await pipe(
-        [rpcQuery.encode()],
-        lp.encode(),
-        stream,
-        lp.decode(),
-        async (source) => await all(source)
-      );
+    const stream = await this.newStream(peer);
 
-      const bytes = new Uint8ArrayList();
-      pipeResponse.forEach((chunk) => {
-        bytes.append(chunk);
-      });
+    const res = await pipe(
+      [rpcQuery.encode()],
+      lp.encode(),
+      stream,
+      lp.decode(),
+      async (source) => await all(source)
+    );
 
-      const decoded = PeerExchangeRPC.decode(bytes).response;
+    console.log({ res });
 
-      if (!decoded) {
-        throw new Error("Failed to decode response");
-      }
+    while (!this._pxResponse) {
+      console.log("waiting for 1 sec", this._pxResponse);
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+    }
```

Comment:
Very hackish :)

If this protocol is more intended to be like filter then this should be reflected in the API.

See https://github.com/waku-org/nwaku/issues/1408#issuecomment-1328444519


My thoughts would be:

1. Need to add a request id to the requests
2. Track that we are waiting for a response for this id in a map
3. Once response is received, check that we are waiting for a response, if so, process response, if not, discard.

In terms of the API, I think it would make most sense to have this API emit events when receiving a response.

For now the event could just be the peer details.

Later, we may want to add more info such as request/response id, peer id of the peer we used to request etc.

Reply:
Until we get @kaiserd to review/agree/implement in nwaku, we can still move forward by using the peer id of the remote peer as a unique id for the request/response.

Actually, maybe this can be the final design too. Not sure.

Reply:
> using the peer id of the remote peer as a unique id for the request/response.
Not sure I fully understand this.
How will we map a request/response with a peer ID as there is no information of the peer/peer id on the response? 
Maybe we could leverage the stream metadata for this but the outbound stream is different from the inbound stream, so would it require to be handled on nwaku side?

Reply:
In a conversation with @fryorcraken, it was concluded that this is currently on hold until the peer-exchange overhaul is implemented on nwaku.
Will open an issue to track the follow up steps required from nwaku side to support js-waku fully.